### DISCUSSION TITLE
What can I do to intercept ws

### DISCUSSION THREAD

**Discussion by zzkkui** [eclipse-theia/theia#9776] [Q&A] [answered]
Hi,
I need to disconnect after inactivity to release resources. So I want to know how to get the inactive state. At present, I have an idea to intercept ws and determine whether it is active based on the information inside, but I have not found a way to intercept it.

If there is currently no method, I think the message parameter should be added in the following method

https://github.com/eclipse-theia/theia/blob/7d3b1d3b33cb565ec07965e5baf10c926006cdfd/packages/core/src/common/messaging/abstract-connection-provider.ts#L51

https://github.com/eclipse-theia/theia/blob/7d3b1d3b33cb565ec07965e5baf10c926006cdfd/packages/core/src/common/messaging/abstract-connection-provider.ts#L115

Thanks

**Comment by msujew** [comment, marked as answer]
Hi @zzkkui thank you for your question,

I assume the resources you want to release are on the backend? Correct me if I'm wrong there. So basically, the `MessagingContribution` will call `terminate` on the web socket once it doesn't answer for 30 seconds. See here:

https://github.com/eclipse-theia/theia/blob/33dea9b48f84eece471357ad271ed2071cc7b012/packages/core/src/node/messaging/messaging-contribution.ts#L109-L117

In order to use that to release your resources, you will have to add your own function to the `onClose` callback of the `WebSocketChannel` which is created for each user. I think the best way to add that `onClose` callback is to create your custom `MessagingContribution` and override the `createChannel` method. Just add your callback to the `WebSocketChannel` before returning it:

https://github.com/eclipse-theia/theia/blob/33dea9b48f84eece471357ad271ed2071cc7b012/packages/core/src/node/messaging/messaging-contribution.ts#L217-L227

**Reply by zzkkui** [reply]
Hi @msujew , Thank you for your answer. The inactive state I am talking about here is similar to that gitpod will automatically shut down after a period of inactivity. I don’t know if I’m accurate
In fact, I want to implement this function, and I guess it is judged by intercepting ws information

like this information
![image](https://user-images.githubusercontent.com/24466870/126978918-092fc7fe-bfa1-4e1e-a26f-f2c1a1d19ef0.png)
or
![image](https://user-images.githubusercontent.com/24466870/126978930-e791370a-508a-4867-80af-4c20fad58321.png)

Thanks any more

**Reply by msujew** [reply]
Oh I see, that's how it is. Alright then, the actual approach isn't that different compared to what I proposed earlier. However, instead of adding your callback to the `onClose` method of the `WebSocketChannel`, you can simply add a callback to `onMessage`, which contains the parsed json data: (At least I think that this should work)

https://github.com/eclipse-theia/theia/blob/65509341ebf2e6c2aedd98c066c9d405a111200f/packages/core/src/common/messaging/web-socket-channel.ts#L118-L122

**Reply by zzkkui** [reply]
Yes, I was also studying here before, but I want to modify and overwrite `onMessage` through extensions instead of directly modifying the source code of theia. Unfortunately, I did not find a way to achieve it through extensions. This makes me very frustrated. 
So I hope to get some help, thank you very much

**Reply by msujew** [reply]
Yeah, you should be easily able to overwrite `onMessage` callback for your own extension without the need to change the Theia source code. See the following example:

```ts
export class CustomMessagingContribution extends MessagingContribution {
    protected createChannel(id: number, socket: ws): WebSocketChannel { 
        const channel = super.createChannel(id, socket);
        channel.onMessage(data => {
            // insert your custom logic in here
        };
        return channel;
    } 
}
```

You would also have to rebind the `MessagingContribution` in your backend module:

```ts
rebind(MessagingContribution).to(CustomMessagingContribution).inSingleton();
```

**Reply by zzkkui** [reply]
Oh !!! Thank you for your answer. I tried to use your method and did get the right to process the message, but there is a difference, I can't get the message in `onMessage`, I get the message in `handleMessage`. It's all because of your help, thank you very much!

following example:

```typescript
protected createChannel(id: number, socket: ws) {
        const channel = super.createChannel(id, socket);
        const handleMessage = channel.handleMessage.bind(channel);
        channel.handleMessage = (message: WebSocketChannel.Message) => {
            handleMessage(message);
            if(message.kind === 'data') {
                console.log(message.content);
            }
        };
        return channel;
    }
```

and then rebind 
```typescript
rebind<MessagingContribution>(MessagingService.Identifier).to(CustomMessagingContribution).inSingletonScope();
```

**Reply by msujew** [reply]
Oh, I see, I didn't notice that the `kind` property of your message is `data` in the example you provided. `onMessage` only handles the `message` kind, but not `data`. Glad you figured that out on your own, good job!
